File: src/main/java/algorithm/baekjoon/_1946_sort.java
package algorithm.baekjoon;

import java.util.*;
import java.io.*;

// https://www.acmicpc.net/problem/1946
/**
 * 풀이
 *
 * 1. 서류 혹은 면접 둘중 하나의 순위 기준으로 오름차순 정렬한다.
 * 2. 정렬 대상값이 아닌 걸 기준으로 나머지 순위를 비교함
 * 3. 이전보다 더 낮은 순위라면 갱신한다 (더 잘한 지원자 임으로)
 *
 */
public class _1946_sort {

    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int T = Integer.parseInt(st.nextToken());

        while (T > 0) {

            st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken());

            // 객체 생성
            List<Applicants> applicants = new ArrayList<>();
            for (int i = 0; i < N; i++) {
                st = new StringTokenizer(br.readLine());
                int a = Integer.parseInt(st.nextToken());
                int b = Integer.parseInt(st.nextToken());

                Applicants ap = new Applicants(a, b);
                applicants.add(ap);
            }

            // 서류 순위 기준으로 오름차 순
            applicants.sort(new Comparator<Applicants>() {
                @Override
                public int compare(Applicants o1, Applicants o2) {
                    return o1.paper - o2.paper;
                }
            });

            System.out.println(applicants);

            // 첫번째는 무조건 선발 (1순위 임으로)
            int answer = 1;

            // 서류 기준 인터뷰 순위
            int maxInterviewCnt = applicants.get(0).interview;
            for (int i = 0; i < applicants.size(); i++) {
                // 해당 서류
                if (maxInterviewCnt > applicants.get(i).interview) {
                    answer ++;
                    maxInterviewCnt = applicants.get(i).interview;
                }
            }
            System.out.println(answer);

            T--;
        }
    }

    public static class Applicants {
        int interview;
        int paper;

        Applicants(int interview, int paper) {
            this.interview = interview;
            this.paper = paper;
        }

        @Override
        public String toString() {
            return "interview : " + interview + " paper : " + paper + "\n";
        }
    }
}

File: src/main/java/algorithmKotlin/baekjoon/_1946_sort.kt
package algorithmKotlin.baekjoon

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

class _1946_sort {

//    companion object {
//        @JvmStatic
//        fun main(args: Array<String>) {
//
//            var br = BufferedReader(InputStreamReader(System.`in`))
//            var st = StringTokenizer(br.readLine())
//
//            var T = st.nextToken().toInt()
//
//            repeat(T) {
//
//                st = StringTokenizer(br.readLine())
//                var N = st.nextToken().toInt()
//
//                var lists = mutableListOf<Pair<Int, Int>>()
//                for (i in 0 until N) {
//
//                    st = StringTokenizer(br.readLine())
//
//                    var a = st.nextToken().toInt()
//                    var b = st.nextToken().toInt()
//
//                    lists.add(Pair(a, b))
//                }
//
//                // 최초 한명은 반드시 뽑힌다.
//                var answer = 1
//
//                // 서류 순위로 오름차순 정렬
//                lists.sortBy { it.first }
//                var maxInterview = lists.get(0).second
//
//                for ((a, interviewRank) in lists) {
//                    if (maxInterview > interviewRank) {
//                        maxInterview = interviewRank
//                        answer ++
//                    }
//                }
//
//                println(answer)
//            }
//        }
//    }

    // 코틀린스러운 버전
    companion object {
        @JvmStatic
        fun main(args: Array<String>) {

            val br = System.`in`.bufferedReader()
            val sb = StringBuilder()

            // 테스트 케이스 수
            val t = br.readLine()!!.toInt()
            repeat(t) {
                // 지원자 수
                val n = br.readLine()!!.toInt()

                // 지원자 리스트를 생성하고 서류 순위 기준으로 정렬
                val applicants = List(n) {
                    br.readLine()!!.split(' ').let { (doc, interview) ->
                        doc.toInt() to interview.toInt()
                    }
                }.sortedBy { it.first }

                // 첫 지원자는 무조건 선발
                var count = 1
                var bestInterview = applicants.first().second

                // 나머지 지원자 중 면접 순위가 더 좋은 경우만 카운트
                for ((_, interview) in applicants.drop(1)) {
                    if (interview < bestInterview) {
                        count++
                        bestInterview = interview
                    }
                }

                sb.appendLine(count)
            }

            print(sb)
        }
    }
}